src/rgw/driver/d4n/rgw_sal_d4n.cc

```diff
@@ -1878,31 +1920,53 @@ int D4NFilterObject::D4NFilterDeleteOp::delete_obj(const DoutPrefixProvider* dpp
 	    ldpp_dout(dpp, 0) << "Failed to delete head object for: " << source->get_key().get_oid() << ", ret=" << ret << dendl;
 	    return ret;
 	  }
-        } else {
-	  ldpp_dout(dpp, 0) << "Failed to delete versioned head object in block directory for: " << source->get_key().get_oid() << ", ret=" << ret << dendl;
-	  return ret;
-	}
+        }
 
-	auto headObj = block;
-	headObj.cacheObj.objName = source->get_name();
-	ret = blockDir->get(dpp, &headObj, y); // retrieve head object
-	if (!block.prevVersion && ret == 0 && headObj.version == version) { // if the latest version matches the provided version and there are no 
-									    // previous versions left, this is the last version of the object
-	  ldpp_dout(dpp, 10) << "D4NFilterObject::D4NFilterDeleteOp::" << __func__ << "(): No previous version found; deleting head object" << dendl;
+        if (!nullDelete) { 
+	  auto headObj = block;
+	  headObj.cacheObj.objName = source->get_name();
+	  ret = blockDir->get(dpp, &headObj, y); // retrieve head object
+	  if (!block.prevVersion && ret == 0) {
+            if (headObj.version == version) { // if the latest version matches the provided version and there are no 
+					      // previous versions left, this is the last version of the object
+	      ldpp_dout(dpp, 10) << "D4NFilterObject::" << __func__ << "(): no previous version found; deleting head object for " << headObj.cacheObj.objName << dendl;
+
+	      if ((ret = blockDir->del(dpp, &headObj, y)) < 0 && ret != -ENOENT) { // delete head object
+		ldpp_dout(dpp, 0) << "Failed to delete head object in block directory for: " << source->get_key().get_oid() << ", ret=" << ret << dendl;
+		return ret;
+	      }
+
+	      if ((ret = source->driver->get_obj_dir()->del(dpp, &headObj.cacheObj, y)) < 0 && ret != -ENOENT) {
```

Comment:
This doesn't exist anymore, in one of my latest commits I have removed the objects being stored in the object directory, as we need to come up with a better data structure to store objects indexed by bucket name. So you can probably leave a place holder here by adding a comment